type_classes.md

```diff
@@ -29,73 +37,179 @@ def double (s : Add a) (x : a) : a :=
 #eval double { add := Nat.add } 10
 -- 20
 
-#eval double { add := Nat.mul } 10
--- 100
-
 #eval double { add := Int.add } 10
 -- 20
 
 # end Ex
 ```
-Note that you can double a natural number `n` by `double { add := Nat.add } n`.
+Note that you can double a natural number or an integer by providing
+the right implementation of those add functions. Notice also that the
+invocation of the add function simply passes x twice and doesn't
+really care what that function does so this also works even though
+this would be unexpected:
+
+```lean
+# namespace Ex
+# structure Add (a : Type) where
+#  add : a -> a -> a
+# def double (s : Add a) (x : a) : a :=
+#  s.add x x
+#eval double { add := Nat.mul } 10  -- 100
+```
+
 Of course, it would be highly cumbersome for users to manually pass the
 implementations around in this way.
 Indeed, it would defeat most of the potential benefits of ad-hoc
 polymorphism.
 
 The main idea behind type classes is to make arguments such as `Add a` implicit,
-and to use a database of user-defined instances to synthesize the desired instances
-automatically through a process known as typeclass resolution. In Lean, by changing
-`structure` to `class` in the example above, the type of `Add.add` becomes
+and to use all these type classes to synthesize the desired `Type` instances
+automatically through a process known as *type class resolution*. In Lean, by changing
+`structure` to `class` in the example above, the type of `Add.add` becomes:
 ```lean
 # namespace Ex
 class Add (a : Type) where
   add : a -> a -> a
 
-#check @Add.add
--- Add.add : {a : Type} → [self : Add a] → a → a → a
+#check Add Nat      -- Add Nat : Type
 # end Ex
 ```
-where the square brackets indicate that the argument of type `Add a` is *instance implicit*,
-i.e. that it should be synthesized using typeclass resolution. This version of
-`add` is the Lean analogue of the Haskell term `add :: Add a => a -> a -> a`.
-Similarly, we can register an instance by
+
+Note that an "instance" of a type class is a `Type` so you can think of
```

Comment:
thanks for clarifying, I was quite confused about this, so I will rework this bit.  The model I had in my head was like a C++ template Foo<T> is like a typeclass and the "instantiation" Foo<int> is like the Lean "instance", but in C++ Foo<int> is still a Type.  But you are saying that the "instance" is is not just the instantiated type "Test Nat" but is also an object of that instantiated type.  The doc is also missing the notion that you can "name" the instance, I think that makes things much more clear so I will add that too.

Reply:
You are totally right, a class operates in the same was as a structure.  So why do we call it a "type class" instead of just a "class" ???

```lean
class Point (α : Type) where
  x : α
  y : α
  deriving Repr

#eval { x := 1, y:= 1 : Point Nat }   -- { x := 1, y := 1 }

structure sPoint (α : Type) where
  x : α
  y : α
  deriving Repr

#eval { x := 1, y:= 1 : sPoint Nat }  -- { x := 1, y := 1 }
```

Reply:
While classes are just structures under the hood in Lean they are meant to be used for polymorphism. For example the type class `Repr` describes the class of types whose values have the capability to be represented. The `Add` type class describes the class of types whose values have the capability to be added heterogeneously and so on. So that's the reason for the naming.